File: rust-mvp/src/commands/skills.rs
//! Skill aliases — discovered skills as first-class palette commands.
//!
//! Port of `src/amplifier_app_newtui/commands/skills.py`.
//!
//! Brian's story #1: `/cranky-old-sam` (and its `shortcut:` alias
//! `/cosam`) must resolve exactly like any built-in before slash input
//! can fall through as a chat turn. Rather than a second lookup table,
//! each discovered skill registers additively into the ONE command
//! registry (ADR-0007: commands are data + callables) — so the palette,
//! help listing, `parse_and_run` dispatch and the unknown-command check
//! all see skills for free.
//!
//! Layering: skills arrive duck-typed (`name` / `description` /
//! `shortcut` accessors, i.e. the `kernel.session_ops.SkillInfo` shape —
//! the [`SkillLike`] trait here, Python's `Protocol`) — this module still
//! imports nothing above `model/`. Handlers invoke the skill through
//! [`CommandContext::load_skill`], the same path the built-in
//! `/skill <name>` takes.

use std::collections::HashSet;
use std::sync::Arc;

use super::registry::{
    CommandContext, CommandGroup, CommandHandler, CommandRegistry, CommandSpec,
};

/// Registration source label for skill-contributed commands.
const SKILL_SOURCE: &str = "skill";

/// What a discovered skill must offer (`session_ops.SkillInfo` shape).
///
/// Python is a `Protocol` with `name` / `description` / `shortcut`
/// string properties; absent values surface as empty strings (the Python
/// call sites coerce with `str(value or "")`).
pub trait SkillLike {
    fn name(&self) -> String;
    fn description(&self) -> String;
    fn shortcut(&self) -> String;
}

fn load_handler(skill_name: &str) -> CommandHandler {
    let skill_name = skill_name.to_string();
    // Alias arguments are not plumbed through load_skill (yet).
    Arc::new(move |ctx: &dyn CommandContext, _args: &str| ctx.load_skill(&skill_name))
}

/// A `skill`-tagged spec for *trigger*, or `None` when the token is not
/// a valid slash trigger (spaces, empty — validator decides).
fn spec_for(trigger: &str, desc: &str, skill_name: &str) -> Option<CommandSpec> {
    CommandSpec::new(
        CommandGroup::During,
        &format!("/{trigger}"),
        desc,
        "skill",
        load_handler(skill_name),
    )
    .ok()
}

/// Palette rows for *skills* that don't collide with *registry*.
///
/// One row per skill name plus one per distinct `shortcut` alias
/// (the alias row names its target so the palette reads as an alias).
/// Collisions with already-registered commands — built-ins or earlier
/// skills — are skipped: first registration wins, never overridden.
pub fn skill_command_specs<S: SkillLike>(
    registry: &CommandRegistry,
    skills: &[S],
) -> Vec<CommandSpec> {
    let mut specs: Vec<CommandSpec> = Vec::new();
    let mut taken: HashSet<String> = registry.names().into_iter().collect();
    for skill in skills {
        let name = skill.name().trim().to_string();
        let joined = skill
            .description()
            .split_whitespace()
            .collect::<Vec<_>>()
            .join(" ");
        let desc = if joined.is_empty() {
            format!("load skill {name}")
        } else {
            joined
        };
        let Some(spec) = spec_for(&name, &desc, &name) else {
            continue;
        };
        if taken.contains(&spec.name) {
            continue;
        }
        taken.insert(spec.name.clone());
        specs.push(spec);
        let shortcut = skill.shortcut().trim().to_string();
        if !shortcut.is_empty() && shortcut != name {
            if let Some(alias) = spec_for(&shortcut, &format!("{name} · {desc}"), &name) {
                if !taken.contains(&alias.name) {
                    taken.insert(alias.name.clone());
                    specs.push(alias);
                }
            }
        }
    }
    specs
}

/// Register *skills* (names + shortcuts) into *registry*; returns the
/// specs actually added — empty when everything was already present.
///
/// Rides the open-registry mechanism (story #2): each row registers as
/// a `skill`-sourced contribution, so `registry.contributions("skill")`
/// lists them and the registry's own collision policy (existing command
/// wins, skip with a log line) backstops the prefilter above.
pub fn register_skill_commands<S: SkillLike>(
    registry: &CommandRegistry,
    skills: &[S],
) -> Vec<CommandSpec> {
    skill_command_specs(registry, skills)
        .into_iter()
        .filter(|spec| {
            registry
                .register_with_source(spec.clone(), SKILL_SOURCE)
                .expect("non-builtin registrations never error")
        })
        .collect()
}

// ---------------------------------------------------------------------------
// Tests — port of tests/test_commands_skills.py (all cases). The fake
// context mirrors tests/conftest.py's FakeCommandContext; `build_registry`
// stands in for commands/builtin.py's seed (not yet ported) with just the
// built-ins these tests collide with (/mode, /status, /skill), using the
// real builtin descriptions and handler behaviors.
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use std::any::Any;
    use std::sync::Mutex;

    use rust_decimal::Decimal;

    use super::*;
    use crate::model::blocks::{BlockIdAllocator, TranscriptBlock};
    use crate::model::queues::{NeedsYouQueue, SteeringQueue};
    use crate::model::trust::DenialLog;
    use crate::model::turn::OutcomeLedger;

    /// Python `_skill(name, description="", shortcut="")` (SimpleNamespace).
    struct TestSkill {
        name: String,
        description: String,
        shortcut: String,
    }

    impl SkillLike for TestSkill {
        fn name(&self) -> String {
            self.name.clone()
        }

        fn description(&self) -> String {
            self.description.clone()
        }

        fn shortcut(&self) -> String {
            self.shortcut.clone()
        }
    }

    fn skill(name: &str, description: &str, shortcut: &str) -> TestSkill {
        TestSkill {
            name: name.to_string(),
            description: description.to_string(),
            shortcut: shortcut.to_string(),
        }
    }

    /// Minimal stand-in for `commands.builtin.build_registry()` — the
    /// builtin module is a later unit; these are the three seeded rows the
    /// pinned tests touch, with their real descriptions and handlers.
    fn build_registry() -> CommandRegistry {
        let mode: CommandHandler = Arc::new(|ctx: &dyn CommandContext, args: &str| {
            if args.trim().is_empty() {
                ctx.cycle_mode();
            } else {
                ctx.set_mode(args.trim());
            }
        });
        let status: CommandHandler =
            Arc::new(|ctx: &dyn CommandContext, _args: &str| ctx.show_status());
        let skill: CommandHandler =
            Arc::new(|ctx: &dyn CommandContext, args: &str| ctx.load_skill(args.trim()));
        CommandRegistry::with_specs([
            CommandSpec::new(
                CommandGroup::During,
                "/mode",
                "cycle or jump posture: chat, plan, brainstorm, build, auto",
                "built-in",
                mode,
            )
            .unwrap()
            .with_key_action("cycle_mode"),
            CommandSpec::new(
                CommandGroup::During,
                "/status",
                "session status: model, mode, messages, cost",
                "built-in",
                status,
            )
            .unwrap(),
            CommandSpec::new(
                CommandGroup::During,
                "/skill",
                "load a skill by name: /skill <name>",
                "skill",
                skill,
            )
            .unwrap(),
        ])
        .expect("seed registry builds")
    }

    /// Port of `tests/conftest.py::FakeCommandContext` — records every
    /// action a command handler takes.
    struct FakeCommandContext {
        ledger: Mutex<OutcomeLedger>,
        denial_log: Mutex<DenialLog>,
        steering: SteeringQueue,
        needs_you: NeedsYouQueue,
        ids: Mutex<BlockIdAllocator>,
        user_lines: Mutex<Vec<String>>,
        blocks: Mutex<Vec<TranscriptBlock>>,
        notices: Mutex<Vec<String>>,
        calls: Mutex<Vec<String>>,
    }

    impl FakeCommandContext {
        fn new() -> Self {
            Self {
                ledger: Mutex::new(OutcomeLedger::new()),
                denial_log: Mutex::new(DenialLog::new()),
                steering: SteeringQueue::new(),
                needs_you: NeedsYouQueue::new(),
                ids: Mutex::new(BlockIdAllocator::new()),
                user_lines: Mutex::new(Vec::new()),
                blocks: Mutex::new(Vec::new()),
                notices: Mutex::new(Vec::new()),
                calls: Mutex::new(Vec::new()),
            }
        }

        fn record(&self, entry: impl Into<String>) {
            self.calls.lock().unwrap().push(entry.into());
        }

        fn user_lines(&self) -> Vec<String> {
            self.user_lines.lock().unwrap().clone()
        }

        fn calls(&self) -> Vec<String> {
            self.calls.lock().unwrap().clone()
        }
    }

    impl CommandContext for FakeCommandContext {
        fn ledger(&self) -> &Mutex<OutcomeLedger> {
            &self.ledger
        }

        fn denial_log(&self) -> &Mutex<DenialLog> {
            &self.denial_log
        }

        fn steering(&self) -> &SteeringQueue {
            &self.steering
        }

        fn needs_you(&self) -> &NeedsYouQueue {
            &self.needs_you
        }

        fn session_cost(&self) -> Decimal {
            Decimal::ZERO
        }

        fn session_short(&self) -> String {
            "a1b2c3".to_string()
        }

        fn bundle_name(&self) -> String {
            "dev-bundle".to_string()
        }

        fn next_block_id(&self) -> String {
            self.ids.lock().unwrap().next_id()
        }

        fn context_usage(&self) -> Box<dyn Any> {
            Box::new(())
        }

        fn approval_tallies(&self) -> Vec<Box<dyn Any>> {
            Vec::new()
        }

        fn overridden_denials(&self) -> Vec<Box<dyn Any>> {
            Vec::new()
        }

        fn mcp_server_stats(&self) -> Vec<Box<dyn Any>> {
            Vec::new()
        }

        fn echo_user_line(&self, text: &str) {
            self.user_lines.lock().unwrap().push(text.to_string());
        }

        fn post_block(&self, block: TranscriptBlock) {
            self.blocks.lock().unwrap().push(block);
        }

        fn show_notice(&self, text: &str) {
            self.notices.lock().unwrap().push(text.to_string());
        }

        fn cycle_mode(&self) {
            self.record("cycle_mode");
        }

        fn set_mode(&self, mode_id: &str) {
            self.record(format!("set_mode:{mode_id}"));
        }

        fn set_theme(&self, name: &str) {
            self.record(format!("set_theme:{name}"));
        }

        fn toggle_lanes(&self) {
            self.record("toggle_lanes");
        }

        fn open_rewind(&self) {
            self.record("open_rewind");
        }

        fn open_permissions(&self) {
            self.record("open_permissions");
        }

        fn manage_directories(&self, kind: &str, args: &str) {
            self.record(format!("manage_directories:{kind}:{args}"));
        }

        fn quit_app(&self) {
            self.record("quit_app");
        }

        fn export_transcript(&self) -> String {
            self.record("export_transcript");
            "exports/a1b2c3-20260101-000000.md".to_string()
        }

        fn copy_answer(&self) -> usize {
            self.record("copy_answer");
            42
        }

        fn about_info(&self) -> (String, String, String, String) {
            self.record("about_info");
            (
                "0.1.0".to_string(),
                "1.2.3".to_string(),
                self.bundle_name(),
                self.session_short(),
            )
        }

        fn show_modes(&self) {
            self.record("show_modes");
        }

        fn set_native_mode(&self, name: Option<&str>) {
            self.record(format!("set_native_mode:{}", name.unwrap_or("None")));
        }

        fn remove_native_mode(&self, name: &str) {
            self.record(format!("remove_native_mode:{name}"));
        }

        fn show_status(&self) {
            self.record("show_status");
        }

        fn show_model(&self, arg: &str) {
            self.record(format!("show_model:{arg}"));
        }

        fn apply_effort(&self, arg: &str) {
            self.record(format!("apply_effort:{arg}"));
        }

        fn compact_context(&self, focus: &str) {
            self.record(format!("compact_context:{focus}"));
        }

        fn clear_context(&self) {
            self.record("clear_context");
        }

        fn show_tools(&self) {
            self.record("show_tools");
        }

        fn show_agents(&self) {
            self.record("show_agents");
        }

        fn show_diff(&self, arg: &str) {
            self.record(format!("show_diff:{arg}"));
        }

        fn show_skills(&self) {
            self.record("show_skills");
        }

        fn load_skill(&self, name: &str) {
            self.record(format!("load_skill:{name}"));
        }

        fn manage_mcp(&self, args: &str) {
            self.record(format!("manage_mcp:{args}"));
        }

        fn load_bundle(&self, args: &str) {
            self.record(format!("load_bundle:{args}"));
        }

        fn manage_config(&self, args: &str) {
            self.record(format!("manage_config:{args}"));
        }

        fn rename_session(&self, name: &str) {
            self.record(format!("rename_session:{name}"));
        }

        fn show_sessions(&self) {
            self.record("show_sessions");
        }

        fn branch_session(&self, name: &str) {
            self.record(format!("branch_session:{name}"));
        }

        fn fork_session(&self, directive: &str) {
            self.record(format!("fork_session:{directive}"));
        }
    }

    fn added_names(added: &[CommandSpec]) -> Vec<&str> {
        added.iter().map(|spec| spec.name.as_str()).collect()
    }

    #[test]
    fn test_registers_skill_and_shortcut_rows() {
        let registry = build_registry();
        let added = register_skill_commands(
            &registry,
            &[skill("cranky-old-sam", "crusty review", "cosam")],
        );
        assert_eq!(added_names(&added), vec!["/cranky-old-sam", "/cosam"]);
        let spec = registry.get("/cranky-old-sam").expect("registered");
        assert_eq!(spec.tag, "skill");
        assert!(spec.desc.contains("crusty review"));
        let alias = registry.get("/cosam").expect("registered");
        assert_eq!(alias.tag, "skill");
        // Alias row names its target.
        assert!(alias.desc.contains("cranky-old-sam"));
    }

    #[test]
    fn test_parse_and_run_resolves_name_and_shortcut() {
        let fake = FakeCommandContext::new();
        let registry = build_registry();
        register_skill_commands(
            &registry,
            &[skill("cranky-old-sam", "crusty review", "cosam")],
        );
        assert!(registry.parse_and_run(&fake, "/cranky-old-sam"));
        assert!(registry.parse_and_run(&fake, "/cosam"));
        // Both routes invoke the skill exactly like `/skill <name>` does.
        assert_eq!(
            fake.calls(),
            vec!["load_skill:cranky-old-sam", "load_skill:cranky-old-sam"]
        );
        assert_eq!(fake.user_lines(), vec!["/cranky-old-sam", "/cosam"]);
    }

    #[test]
    fn test_skips_collisions_with_existing_commands() {
        let fake = FakeCommandContext::new();
        let registry = build_registry();
        let added = register_skill_commands(
            &registry,
            &[
                skill("status", "shadows a built-in", ""), // /status is built-in
                skill("review", "fine", "skill"),          // /skill is built-in
            ],
        );
        assert_eq!(added_names(&added), vec!["/review"]);
        // The built-in survives untouched.
        registry.parse_and_run(&fake, "/status");
        assert_eq!(fake.calls(), vec!["show_status"]);
    }

    #[test]
    fn test_skips_tokens_that_are_not_slash_triggers() {
        let registry = build_registry();
        let added = register_skill_commands(
            &registry,
            &[
                skill("bad name with spaces", "", ""),
                skill("", "", ""),
                skill("ok", "", ""),
            ],
        );
        assert_eq!(added_names(&added), vec!["/ok"]);
    }

    #[test]
    fn test_shortcut_equal_to_name_registers_once() {
        let registry = build_registry();
        let added = register_skill_commands(&registry, &[skill("simplify", "cut", "simplify")]);
        assert_eq!(added_names(&added), vec!["/simplify"]);
    }

    #[test]
    fn test_empty_description_gets_a_default() {
        let registry = build_registry();
        register_skill_commands(&registry, &[skill("terse", "", "")]);
        let spec = registry.get("/terse").expect("registered");
        assert!(!spec.desc.trim().is_empty());
    }

    #[test]
    fn test_registering_twice_is_idempotent() {
        let registry = build_registry();
        let skills = [skill("cranky-old-sam", "crusty review", "cosam")];
        register_skill_commands(&registry, &skills);
        assert_eq!(register_skill_commands(&registry, &skills), Vec::new());
    }

    #[test]
    fn test_skill_rows_are_skill_sourced_contributions() {
        // Story #2: skills ride the open-registry mechanism — their rows are
        // 'skill'-sourced contributions, unregisterable as a group, distinct
        // from the seeded built-ins.
        let registry = build_registry();
        let added = register_skill_commands(
            &registry,
            &[skill("cranky-old-sam", "crusty review", "cosam")],
        );
        assert_eq!(registry.contributions("skill"), added);
        assert_eq!(registry.source_of("/cosam").as_deref(), Some("skill"));
        assert_eq!(registry.source_of("/mode").as_deref(), Some("builtin"));
        assert!(registry.unregister("/cosam").unwrap());
        assert!(registry.get("/cosam").is_none());
        assert!(registry.get("/cranky-old-sam").is_some());
    }
}

File: src/amplifier_app_tui/commands/skills.py
"""Skill aliases — discovered skills as first-class palette commands.

Brian's story #1: ``/cranky-old-sam`` (and its ``shortcut:`` alias
``/cosam``) must resolve exactly like any built-in before slash input
can fall through as a chat turn. Rather than a second lookup table,
each discovered skill registers additively into the ONE command
registry (ADR-0007: commands are data + callables) — so the palette,
help listing, ``parse_and_run`` dispatch and the unknown-command check
all see skills for free.

Layering: skills arrive duck-typed (``name`` / ``description`` /
``shortcut`` attributes, i.e. ``kernel.session_ops.SkillInfo``) — this
package still imports nothing above ``model/``. Handlers invoke the
skill through :meth:`CommandContext.load_skill`, the same path the
built-in ``/skill <name>`` takes.

Compliance B2 (alias fixture parity, 2026-08-02 audit follow-up) added
three things on top of the story #1 baseline, all in this one file so
the registry stays the single source of truth for both the CLI-style
headless resolution path (``CommandRegistry.parse_and_run`` against a
plain ``CommandContext`` — see ``tests/test_commands_skills.py``) and
the interactive TUI path (the real app over Textual — see
``tests/test_flow_skill_aliases.py``); both import the same fixture
from ``tests/test_skill_alias_fixture.py``:

1. **Argument forwarding** (judgment call, recorded here since the
   compliance brief left it open): text after a skill's name-as-command
   or its alias is now forwarded to :meth:`CommandContext.load_skill`
   using the EXACT same concatenation the built-in ``/skill <name>
   <rest>`` already uses (``ctx.load_skill(f"{name} {rest}")``). This
   was chosen over inventing a new argument channel because it makes
   all three spellings — ``/skill cranky-old-sam draft it``,
   ``/cranky-old-sam draft it``, ``/cosam draft it`` — issue byte
   identical ``ctx.load_skill`` calls (CLI/TUI parity, AC1), and because
   this package does not own (and should not guess at) the mounted
   skills tool's own argument contract.
2. **Collision diagnostics** (:class:`AliasCollision`, returned by
   :func:`plan_skill_commands` / :func:`register_skill_commands_reporting`):
   a skill name or shortcut that is already taken — by a built-in or by
   an earlier skill/alias in the same discovery pass — used to vanish
   silently (first registration wins, nothing said). It is now reported
   deterministically so the caller can surface it at configuration load
   (``ui.app.TuiApp._register_skill_commands``, AC4) instead of a quiet
   skip.
3. **Nearby suggestions** for an unrecognized slash command/alias are
   :meth:`~.registry.CommandRegistry.suggest` (AC3), a small addition to
   the registry itself (every registered trigger, skill or built-in,
   benefits — not just skills).
"""

from __future__ import annotations

from collections.abc import Iterable, Sequence
from typing import Any, NamedTuple, Protocol

from pydantic import ValidationError

from ..model.blocks import Segment
from .registry import CommandContext, CommandRegistry, CommandSpec


class SkillLike(Protocol):
    """What a discovered skill must offer (``session_ops.SkillInfo`` shape)."""

    @property
    def name(self) -> str: ...
    @property
    def description(self) -> str: ...
    @property
    def shortcut(self) -> str: ...


class AliasCollision(NamedTuple):
    """One slash trigger a discovered skill wanted but couldn't have.

    Computed by :func:`plan_skill_commands`: the registry's own collision
    policy (first registration wins) already decided the OUTCOME — this
    just names it instead of leaving it a silent skip (AC4).
    """

    trigger: str
    """The slash trigger that collided, e.g. ``/cosam`` or ``/status``."""
    skill: str
    """The skill that wanted *trigger* and did not get it."""
    owner: str
    """Who already holds *trigger*: ``built-in``, ``skill`` (claimed by a
    prior call against this registry), or the specific skill name that
    claimed it earlier in THIS discovery pass."""


class SkillPlan(NamedTuple):
    """What :func:`plan_skill_commands` would register, plus any
    collisions found while deciding that (AC4)."""

    specs: tuple[CommandSpec, ...]
    collisions: tuple[AliasCollision, ...]


def _load_handler(skill_name: str) -> Any:
    def handler(ctx: CommandContext, args: str) -> None:
        # Forward trailing text exactly like the built-in ``/skill <name>
        # <rest>`` path (``_cmd_skill`` in commands/builtin.py) already
        # does — see the module docstring's judgment-call note (AC1).
        rest = args.strip()
        target = f"{skill_name} {rest}" if rest else skill_name
        ctx.load_skill(target)

    return handler


def _spec_for(trigger: str, desc: str, skill_name: str) -> CommandSpec | None:
    """A ``skill``-tagged spec for *trigger*, or ``None`` when the token
    is not a valid slash trigger (spaces, empty — validator decides)."""
    try:
        return CommandSpec(
            group="During",
            name=f"/{trigger}",
            desc=desc,
            tag="skill",
            handler=_load_handler(skill_name),
        )
    except ValidationError:
        return None


def plan_skill_commands(registry: CommandRegistry, skills: Iterable[SkillLike]) -> SkillPlan:
    """Compute the palette rows *skills* would add to *registry*, and any
    alias collisions found along the way (AC4).

    One row per skill name plus one per distinct ``shortcut`` alias (the
    alias row names its target; the canonical row now also names ITS
    alias, so autocomplete/help show both directions — AC2). A trigger
    already claimed — by a built-in, by an earlier skill in *skills*, or
    by a previous call against this same *registry* — is never
    double-registered (first registration wins, unchanged from the
    story #1 baseline); it is instead recorded as an :class:`AliasCollision`
    so the caller can choose to surface it rather than stay silent.

    A skill whose canonical name collides skips its alias too (matching
    the pre-B2 behavior exactly) — there is no useful "half a skill"
    registration.
    """
    specs: list[CommandSpec] = []
    collisions: list[AliasCollision] = []
    # trigger -> who holds it. Seeded from the registry's CURRENT state
    # (built-ins, plus anything a previous discovery pass already added)
    # using each existing spec's own display tag ("built-in"/"skill") —
    # the same word already shown in the palette's dimmer tag column.
    owner: dict[str, str] = {}
    for existing_name in registry.names:
        existing_spec = registry.get(existing_name)
        owner[existing_name] = existing_spec.tag if existing_spec is not None else "built-in"

    for skill in skills:
        name = str(skill.name or "").strip()
        desc = " ".join(str(skill.description or "").split()) or f"load skill {name}"
        shortcut = str(skill.shortcut or "").strip()
        has_alias = bool(shortcut) and shortcut != name
        canonical_desc = f"{desc} · alias /{shortcut}" if has_alias else desc

        spec = _spec_for(name, canonical_desc, name)
        if spec is None:
            continue  # not a valid slash trigger — nothing to collide with
        if spec.name in owner:
            collisions.append(AliasCollision(trigger=spec.name, skill=name, owner=owner[spec.name]))
            continue  # canonical name lost the trigger: alias isn't attempted either
        specs.append(spec)
        owner[spec.name] = name

        if not has_alias:
            continue
        alias_spec = _spec_for(shortcut, f"{name} · {desc}", name)
        if alias_spec is None:
            continue
        if alias_spec.name in owner:
            collisions.append(
                AliasCollision(trigger=alias_spec.name, skill=name, owner=owner[alias_spec.name])
            )
            continue
        specs.append(alias_spec)
        owner[alias_spec.name] = name

    return SkillPlan(specs=tuple(specs), collisions=tuple(collisions))


def skill_command_specs(
    registry: CommandRegistry, skills: Iterable[SkillLike]
) -> tuple[CommandSpec, ...]:
    """Palette rows for *skills* that don't collide with *registry*.

    Thin view over :func:`plan_skill_commands` for callers that only
    want the specs (collision-blind); kept for back-compat with the
    story #1 API.
    """
    return plan_skill_commands(registry, skills).specs


def alias_collision_spans(collisions: Sequence[AliasCollision]) -> tuple[Segment, ...]:
    """A rich diagnostic listing for skill alias collisions (AC4): one
    row per trigger a skill wanted but couldn't claim, naming who already
    holds it. Empty when there is nothing to report. Textual-free (plain
    :class:`~amplifier_app_tui.model.blocks.Segment` spans posted as an
    ``Answer`` block by the caller) — matches the house style of
    ``ui/session_ops_view.py``'s other diagnostic listings
    (``/status``, ``/skills``, ...).
    """
    if not collisions:
        return ()
    count = len(collisions)
    noun = "collision" if count == 1 else "collisions"
    spans: list[Segment] = [
        Segment(text="· ", style_token="blue"),
        Segment(text="Skill aliases", style_token="bright", bold=True),
        Segment(
            text=f"  {count} {noun} · first registration wins\n",
            style_token="dim",
        ),
    ]
    width = max(len(collision.trigger) for collision in collisions)
    for collision in collisions:
        spans.append(Segment(text=f"  {collision.trigger.ljust(width)}  ", style_token="orange"))
        spans.append(
            Segment(
                text=f"wanted by {collision.skill} · already claimed by {collision.owner}\n",
                style_token="dim",
            )
        )
    return tuple(spans)


def register_skill_commands_reporting(
    registry: CommandRegistry, skills: Iterable[SkillLike]
) -> SkillPlan:
    """Register *skills* (names + shortcuts) into *registry*; returns
    both the specs actually added AND any alias collisions found (AC4).

    Rides the open-registry mechanism (story #2): each row registers as
    a ``skill``-sourced contribution, so ``registry.contributions("skill")``
    lists them and the registry's own collision policy (existing command
    wins, skip with a log line) backstops the prefilter in
    :func:`plan_skill_commands`.
    """
    plan = plan_skill_commands(registry, skills)
    added = tuple(spec for spec in plan.specs if registry.register(spec, source="skill"))
    return SkillPlan(specs=added, collisions=plan.collisions)


def register_skill_commands(
    registry: CommandRegistry, skills: Iterable[SkillLike]
) -> tuple[CommandSpec, ...]:
    """Register *skills* (names + shortcuts) into *registry*; returns the
    specs actually added — ``()`` when everything was already present.

    Back-compat wrapper over :func:`register_skill_commands_reporting`
    for callers that don't need collision diagnostics; the TUI boot path
    (AC4) calls the reporting version directly instead.
    """
    return register_skill_commands_reporting(registry, skills).specs


__all__ = [
    "AliasCollision",
    "SkillLike",
    "SkillPlan",
    "alias_collision_spans",
    "plan_skill_commands",
    "register_skill_commands",
    "register_skill_commands_reporting",
    "skill_command_specs",
]
